src/vm-vcpu/src/vm.rs

```diff
@@ -113,36 +167,97 @@ impl Default for VmRunState {
 }
 
 impl<EH: 'static + ExitHandler + Send> KvmVm<EH> {
-    /// Create a new `KvmVm`.
-    pub fn new<M: GuestMemory>(
+    // Helper function for creating a default VM.
+    // This is needed as the same initializations needs to happen when creating
+    // a fresh vm as well as a vm from a previously saved state.
+    fn create_vm<M: GuestMemory>(
         kvm: &Kvm,
-        vm_state: VmState,
-        guest_memory: &M,
+        config: VmConfig,
         exit_handler: EH,
+        guest_memory: &M,
     ) -> Result<Self> {
         let vm_fd = Arc::new(kvm.create_vm().map_err(Error::CreateVm)?);
-
         let vcpu_run_state = Arc::new(VcpuRunState::default());
 
         let vm = KvmVm {
-            vcpu_barrier: Arc::new(Barrier::new(vm_state.num_vcpus as usize)),
-            state: vm_state,
+            vcpu_barrier: Arc::new(Barrier::new(config.num_vcpus as usize)),
+            config,
             fd: vm_fd,
             vcpus: Vec::new(),
             vcpu_handles: Vec::new(),
             exit_handler,
             vcpu_run_state,
         };
-
         vm.configure_memory_regions(guest_memory, kvm)?;
 
+        Ok(vm)
+    }
+
+    /// Create a new `KvmVm`.
+    pub fn new<M: GuestMemory>(
+        kvm: &Kvm,
+        vm_config: VmConfig,
+        guest_memory: &M,
+        exit_handler: EH,
+        bus: Arc<Mutex<IoManager>>,
+    ) -> Result<Self> {
+        let vcpus_config = vm_config.vcpus_config.clone();
+        let mut vm = Self::create_vm(kvm, vm_config, exit_handler, guest_memory)?;
+
         #[cfg(target_arch = "x86_64")]
-        MpTable::new(vm.state.num_vcpus)?.write(guest_memory)?;
+        MpTable::new(vm.config.num_vcpus)?.write(guest_memory)?;
 
-        // TODO: reuse this for setting up the GIC.
         #[cfg(target_arch = "x86_64")]
         vm.setup_irq_controller()?;
 
+        vm.create_vcpus(bus, vcpus_config, guest_memory)?;
+
+        Ok(vm)
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    fn set_state(&mut self, state: VmState) -> Result<()> {
```

Comment:
Similar to `KvmVcpu` https://github.com/rust-vmm/vmm-reference/pull/197/files#r766628831